waypoint/services/nats_service.py

```diff
@@ -0,0 +1,148 @@
+        }
+        if NATS_CREDS_FILE:
+            connect_kwargs["user_credentials"] = NATS_CREDS_FILE
+        else:
+            logger.warning("No NATS credentials file found, assuming local development")
+        nats_client: NATS = await nats.connect(**connect_kwargs)
+
+    except (ErrConnectionClosed, ErrTimeout, ErrNoServers) as e:
+        logger.error(f"Error connecting to NATS server: {e}")
+        raise e
+    logger.debug("Connected to NATS server")
+
+    jetstream: JetStreamContext = nats_client.jetstream()
+    logger.debug("Yielding JetStream context...")
+    yield jetstream
+
+    logger.debug("Closing NATS connection...")
+    await nats_client.close()
+    logger.debug("NATS connection closed")
+
+
+class NatsEventsProcessor:
+    """
+    Class to handle processing of NATS events. Calling the process_events method will
+    subscribe to the NATS server and return an async generator that will yield events
+    """
+
+    def __init__(self, jetstream: JetStreamContext):
+        self.js_context: JetStreamContext = jetstream
+
+    async def _subscribe(
+        self, group_id: str, wallet_id: str
+    ) -> JetStreamContext.PullSubscription:
+        try:
+            logger.debug("Subscribing to JetStream...")
+            if group_id:
+
+                logger.trace("Tenant-admin call got group_id: {}", group_id)
+                subscribe_kwargs = {
+                    "subject": f"{NATS_SUBJECT}.{group_id}.{wallet_id}",
+                    "stream": NATS_STREAM,
+                }
+            else:
+                logger.trace("Tenant call got no group_id")
+                subscribe_kwargs = {
+                    "subject": f"{NATS_SUBJECT}.*.{wallet_id}",
+                    "stream": NATS_STREAM,
+                }
+            subscription = await self.js_context.pull_subscribe(**subscribe_kwargs)
+
+            return subscription
+
+        except BadSubscriptionError as e:
+            logger.error(f"BadSubscriptionError subscribing to NATS: {e}")
+            raise
+        except Error as e:
+            logger.error(f"Error subscribing to NATS: {e}")
+            raise
+        except Exception:
+            logger.exception("Unknown error subscribing to NATS")
+            raise
+
+    @asynccontextmanager
+    async def process_events(
+        self,
+        group_id: str,
+        wallet_id: str,
+        topic: str,
+        stop_event: asyncio.Event,
+        duration: int = 10,
+    ):
+        logger.debug(
+            f"Processing events for group {group_id} and wallet {wallet_id} on topic {topic}"
+        )
+
+        subscription = await self._subscribe(group_id=group_id, wallet_id=wallet_id)
+
+        async def event_generator():
+            end_time = time.time() + duration
+            while not stop_event.is_set():
+                remaining_time = remaining_time = end_time - time.time()
+                logger.trace(f"remaining_time: {remaining_time}")
```

Comment:
Important note: don't use f-string interpolation in logs. Because, trace logs are not evaluated when log level is set to info, but if you use f-string interpolation, then that string is evaluated every time this line gets passed over. if you use the {} interpolation, then it's lazily evaluated, and there's no cost to trace lines when log level is not at trace.

In general, you can use f-string for error logs, since then it doesn't really matter. Those lines are basically always logged in any case. But for trace/debug/info, etc, be sure not to use f-interpolation, since there is unnecessary cost to using it.

So ... please search for `f"` to find all the f-string interpolation used in this PR, and replace with the {} pattern:
```py
logger.trace("remaining_time: {}", remaining_time)
```